### ISSUE TITLE
Adding more detail to the Handshake

### ISSUE THREAD

**Issue by pandafinity** [SocketCluster/socketcluster#120]
Hi :)

I wanted to ask if there is a way to define more specific outcomes for a Handshake - specifying whether the handshake was denied by the server or simply terminated by the client?  We chatted about the '1006' error..

I also wondered if there is a way to have the handshake as a callback - so you can base the rest of your communication on the status of that handshake?

I appreciate there are many reasons to have an implied handshake and simply say 'on connection' but from a logically approach, acknowledgement of all parties wanting to 'talk' is a key step.

Any thoughts would be happily received :)

Thanks.

**Comment by jondubois** [comment]
@pandafinity Ok, I looked into this more deeply.
MIDDLEWARE_HANDSHAKE middleware functions are invoked by the WS module, so it occurs at the WebSocket protocol level. Basically, this is outside of the control of SC.

It looks like the WebSocket procotol does not offer a way to capture the reason why the connection handshake was rejected. See this issue (this is for a different project) https://github.com/theturtle32/WebSocket-Node/issues/46.

That said, it turns out that the ```1006``` status code is not so useless. If you intentionally abort the connection like this:

```js
var socket = socketCluster.connect(...);
socket.disconnect();
```

The 'connectAbort' event on the client will pass error code ```1000``` - But if the connection abort was caused because of a blocked handshake, then the error code will be ```1006```.
So basically ```1006``` means that the connection was refused - This could be either because the handshake was intentionally blocked by the server, that there was a network failure which prevented the connection from being established or that the the client was not allowed to connect because it did not meet origin requirements.

So basically there is nothing we can do about it...
You just have to try to work with the fact that there can only be two states; 1000 (intentional client-initiated abort) and 1006 (non-intentional abort).

**Comment by pandafinity** [comment]
Thanks for digging deeper, knowing the states does help.

I will try and see if I can make an async Handshake function, as I would like all my other SC communication to be done once a valid handshake has been accepted.

**Comment by jondubois** [comment]
Note that the 'connect' event on the client will only trigger once the handshake has been accepted.

**Comment by jondubois** [comment]
As discussed, there is a limitation of the WebSocket protocol itself, hopefully this discussion cleared up things a bit.
